### DISCUSSION TITLE
After Upgrade from Rails 4 to Rails 5 getting error NameError: uninitialized constant Sidekiq::Extensions::DelayedModel

### DISCUSSION THREAD

**Discussion by sonugaur** [sidekiq/sidekiq#5841] [Q&A]
I am working on Rails on rails application that has a version
```
ruby "2.6.6"
gem 'rails', '4.2.11'
gem 'sidekiq', '~> 4.2', '>= 4.2.10'
    
```
Now I am upgrading it to Rails 5 with version
```
ruby "2.7.6"
gem 'rails', '5.1.7'
gem 'sidekiq', '~> 5.2', '>= 5.2.7'
```

I have the following code in sidekiq file **config/initializers/sidekiq.rb**
```
require 'sidekiq'
require 'sidekiq/web'
require 'sidekiq/extensions/action_mailer'
require 'sidekiq/extensions/active_record'
require 'sidekiq/extensions/generic_proxy'
require 'logger'
Sidekiq.configure_server do |config|
  config.redis = { url: ENV['REDISCLOUD_URL'] }
  Rails.logger.formatter = Sidekiq::Logging.logger.formatter
end

Sidekiq.configure_client do |config|
  config.redis = { url: ENV["REDISCLOUD_URL"]}
end

ENV['REDIS_PROVIDER'] = ENV['REDISCLOUD_URL'] || '127.0.0.1:6379'
Sidekiq::Extensions.enable_delay!
```

But I am getting an error on **staging environment** `NameError: uninitialized constant Sidekiq::Extensions::DelayedModel` when any `.delay` method called from a specific code file or from callbacks (existing code)

I have tried to change `.delay` in some parts of the code to `perform_aync` but getting the same error 
And also I find it weird that when I retry all the jobs manually from the staging **rails console** or staging **sidekiq web** the job got executed without any error.

On the Local machine, everything is working fine even though I connected the staging environment on local and it also looks fine

Please refer below screenshot

`before retry`

![Screenshot 2023-03-22 at 10 33 02 PM](https://user-images.githubusercontent.com/30953932/226986509-d85113fa-4ed7-4914-8f64-54485d7c9b84.png)

`after manual retry`

![Screenshot 2023-03-22 at 10 42 11 PM](https://user-images.githubusercontent.com/30953932/226986703-45cee613-761a-45cf-92e3-5a9b3ad65dcf.png)

**Comment by mperham** [comment]
I'm glad to see you upgrading but Sidekiq 4 and 5 are no longer supported. Review the upgrade notes for each major version in `docs/`

**Reply by sonugaur** [reply]
Thanks for your quick response.
So `.delay` will be available if I upgrade to the latest version ??

**Reply by sonugaur** [reply]
Okay so now I have upgraded sidekiq with version 

```
gem 'sidekiq', '~> 6.0'
gem 'redis', '~> 4.1.0'
```

But I am getting the same error, Please help to suggest.

**Reply by mperham** [reply]
I'm not going to walk you through an upgrade. Read the docs and follow the directions. Upgrade ONE major version at a time.

**Reply by sonugaur** [reply]
Yes, I have followed the doc's instructions and made the changes accordingly.
From the last 4-5 days I have tried each and every solution to resolve this error but nothing works. That's why I came here maybe someone who can help in the discussion thread.

I have also made some dirty hacks like below which is executed every 10 min.

```
namespace :retry_sidekiq_job do
  desc "Retry Jobs"
  task add: :environment do
    require 'sidekiq/api'
    Sidekiq::RetrySet.new.retry_all
  end
end

```

**Comment by mperham** [comment]
I'm not sure of the problem: `S::E::DelayedModel` is defined in `sidekiq/extensions/active_record` so NameError makes no sense.

Replace your initializer code with this:

```ruby
ENV['REDIS_PROVIDER'] = 'REDISCLOUD_URL' if ENV['REDISCLOUD_URL']
Sidekiq.configure_server do |config|
  Rails.logger.formatter = Sidekiq::Logging.logger.formatter
end

Sidekiq::Extensions.enable_delay!
```
